MaxEnt.py

```diff
@@ -1,181 +1,122 @@
+    """
 
-    for pair in learned:
-        print(f"{pair[0]:<20}: {pair[1]:<10}")
+    def __init__(self, constraints, weights):
+        self.const = constraints
+        self.w = weights
 
-    return dict(learned)
+    def sorted_by_weights(self):
+        """Displays constraints from highest to lowest weights.
 
+        Returns:
+            list[tuple[str, float]]: a list of (constraint, weight) tuple sorted from highest to lowest weighted.
+        """
+        name_weight_pair = {}
+        for cn, cw in zip(self.const, self.w.squeeze()):
+            name_weight_pair[cn] = np.round(cw, 3)
 
-def predict_prob(vio, cand):
-    with open("predictions.csv", "w", newline="") as file:
-        writer = csv.writer(file)
-        writer.writerow(["UR/Candidate", "Probability"])
-        for t in range(len(vio)):
-            P = compute_probabilities(vio[t], w).squeeze()
-            writer.writerow([urs[t]])
+        sorted_pairs = sorted(
+            name_weight_pair.items(), key=lambda x: x[1], reverse=True
+        )
 
-            for c in range(len(cand[t])):
-                writer.writerow([cand[t][c], f"{round(P[c]*100, 1)}%"])
+        return sorted_pairs
+
+    def compute_probabilities(self, violations):
+        """Predicts probabilities for a particular violation profile.
+
+        Args:
+            violations (np.array[np.array[int]]): the violation profiles of a list of tableaux.
+
+        Returns:
+            np.array[np.array[float]]: predicts probability for each candidate of given URs.
+        """
+        harmonies = np.dot(violations, self.w)
+        exp = np.exp(-harmonies)
+        marginalization = np.sum(exp, axis=0)
+        return exp / marginalization
+
+    def SGD_learn(self, candidates, violations, srs, *regs, iters=10000, eta=0.05):
+        """Updates constraint weights using SGD.
+
+        Args:
+            candidates (np.array[np.array[str]]): candidates considered for each UR.
+            violations (np.array[np.array[int]]): the violation profiles of a list of tableaux.
+            srs (np.array[np.array[float]]): probability of a surface form observed for a given UR.
+            *regs: a variable number of regularization terms. Can be None.
+            iters (int): how many iterations to perform. Defaults to 10000.
+            eta (float): learning rate. Defaults to 0.05.
+
+        Returns:
+            _type_: _description_
+        """
+        # Initialization: begins with the weights that are currently stored.
+        w = self.w
+
+        # Keeps a history of the change in weights.
+        history = [w.squeeze()]
+
+        # Passes through all the training data in each iteration,
+        for _ in range(iters):
+            # Generates a random array of indices of the shape (1, length |URs|).
+            training_order = np.random.choice(
+                a=range(len(violations)), size=len(violations), replace=False
+            )
```

Comment:
I also wonder if for memory efficiency, we can create generators instead of the entire list, especially if we end up working with larger and larger datasets!